File: year-2023/script-language/assignment-09/ex_01.pl
#!/usr/bin/perl
# Hlib-Oleksandr Suliz, Script Language, group no.2

use warnings FATAL => 'all';
use strict;
use Cwd qw(abs_path);
use File::Basename qw(dirname);
require './number_checker.pm';

{
    sub change_name {
        my $name = lc($_[0]);
        $name =~ s/(\w+)/\u$1/g;
        return $name;
    }

    sub print_error {
        $_[2] =~ s/^\s+|\s+$//g;
        printf STDERR "ERROR at line %d (%s): '%s'\n", $_[1], $_[0], $_[2];
    }

    my @files = ();

    foreach my $file (@ARGV) {
        if (-e $file) {
            push(@files, $file);
        }
        else {
            print_error("Main script", 0, "File '$file' does not exist.");
        }
    }

    foreach my $file (@files) {
        my $line_nr = 1;
        my %grades = ();

        open my $fh, '<', $file or die "ERROR: Can't open $file: $!";

        while (<$fh>) {
            my @new_line = split(' ', $_);

            if (scalar @new_line < 3) {
                print_error($file, $line_nr, $_);
            }
            else {
                my $name = change_name($new_line[0]) . " " . change_name($new_line[1]);
                my $value = NumberChecker::check_grade($new_line[2]);

                if ($value != -1) {
                    if (exists($grades{$name})) {
                        @{$grades{$name}}[0] += $value;
                        push(@{$grades{$name}}, $new_line[2]);
                    }
                    else {
                        $grades{$name} = [ 0.0 ];
                        push(@{$grades{$name}}, $value);
                        push(@{$grades{$name}}, $new_line[2]);
                    }
                }
                else {
                    print_error($file, $line_nr, $_);
                }
            }

            $line_nr++;
        }

        $file =~ s{\.[^.]+$}{};
        $file = $file . '.oceny';
        my $file_name = $file;

        open my $of_fh, '>', $file_name or die "ERROR: Can't open $file_name: $!";

        my $all_avg = 0;
        foreach my $k (sort keys %grades) {
            printf $of_fh "%-20s: %s\n", $k, join(' ', @{$grades{$k}}[1 .. $#{$grades{$k}}]);
            $all_avg += @{$grades{$k}}[0] / ($#{$grades{$k}} - 1);
        }

        my $keys_count = scalar keys %grades;
        $all_avg = $all_avg / $keys_count;
        $all_avg = sprintf("%.2f", $all_avg);

        printf $of_fh "\nAverage of all: %s\n", $all_avg;

        close($of_fh);
        close($fh);
    }
}

File: year-2023/script-language/assignment-12/ex_01.py
#!/usr/bin/python3
# Hlib-Oleksandr Suliz, Script Language, group no.2

import os
import re
import sys


def print_error(file, line_nr, line):
    print('ERROR at line ' + str(line_nr) + ' (' + str(file) + '): ' + str(line))


def change_name(name):
    return name.capitalize()


def get_grade(grade):
    if re.search("^[+-]?\d*[.]?\d*$", grade) or re.search("^\d*[.]?\d*[+-]?$", grade):
        if re.search("^[+]+\d*[.]?\d*$", grade):
            value = float(re.split(r'^[+]', grade)[1])

            if value == 0 or (value >= 2 and value <= 5):
                value += 0.25
                return value
            else:
                return -1

        elif re.search("^\d*[.]?\d*[+]+$", grade):
            value = float(re.split(r'[+]$', grade)[0])

            if (value == 0 or (value >= 2 and value <= 5)):
                value += 0.25
                return value
            else:
                return -1

        elif re.search("^[-]+\d*[.]?\d*$", grade):
            value = float(re.split(r'^[-]', grade)[1])

            if value == 0 or (value >= 2.25 and value <= 5):
                value -= 0.25
                return value
            else:
                return -1

        elif re.search("^\d*[.]?\d*[-]+$", grade):
            value = float(re.split(r'[-]$', grade)[0])

            if (value == 0 or (value >= 2.25 and value <= 5)):
                value -= 0.25
                return value
            else:
                return -1

        elif re.search("^\d*[.]?\d*$", grade):
            value = float(grade)
            if value == 0 or (value >= 2 and value <= 5):
                return value
            else:
                return -1

    else:
        return -1


if __name__ == '__main__':
    for i in range(1, len(sys.argv)):
        with open(sys.argv[i], 'r') as my_file:
            line_nr = 1
            grades = {}
            for line in my_file:
                split_line = line.split()

                if len(split_line) < 3:
                    print_error(sys.argv[i], line_nr, line)
                else:
                    name = change_name(split_line[0]) + " " + change_name(split_line[1])
                    grade = get_grade(split_line[2])

                    if grade != -1:
                        if name not in grades:
                            grades[name] = ['0.0']

                        grade_temp = str(float(grades[name][0]) + float(grade))
                        grades[name][0] = grade_temp
                        grades[name].append(split_line[2])

                    else:
                        print_error(sys.argv[i], line_nr, line)

                line_nr += 1

            file_name = os.path.splitext(os.path.basename(sys.argv[i]))[0]
            file_name = file_name + ".oceny"

            new_file = open(file_name, 'w')

            g_avg = 0.0
            for key in sorted(grades):
                new_file.write(key + ": ")

                for i in range(1, len(grades[key])):
                    new_file.write(str(grades[key][i]) + " ")

                avg = round(float(grades[key][0]) / (len(grades[key]) - 1), 2)
                g_avg += avg
                new_file.write(": " + str(avg) + "\n")

            new_file.write("\nAverage of all grades: " + str(round(g_avg / len(grades), 2)))
